File: app/src/main/java/rod/bailey/trafficatnsw/hazard/DownloadHazardFileTask.java
package rod.bailey.trafficatnsw.hazard;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.AsyncTask;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;

import rod.bailey.trafficatnsw.util.ConfigSingleton;
import rod.bailey.trafficatnsw.ui.ListWithEmptyMessage;
import rod.bailey.trafficatnsw.util.MLog;

/**
 * Created by rodbailey on 22/7/17.
 */
public class DownloadHazardFileTask extends AsyncTask<Void, Void, Boolean> {
    private static final String TAG = DownloadHazardFileTask.class.getSimpleName();
    private final Context ctx;
    private final ListWithEmptyMessage listWithEmptyMessage;
    private final HazardListMode mode;
    private ProgressDialog dialog;

    public DownloadHazardFileTask(Context ctx, ListWithEmptyMessage listWithEmptyMessage, HazardListMode mode) {
        this.ctx = ctx;
        this.listWithEmptyMessage = listWithEmptyMessage;
        this.mode = mode;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();

        dialog = new ProgressDialog(ctx);
        dialog.setMessage("Loading incidents...");
        dialog.setCancelable(false);
        dialog.setIndeterminate(true);
        dialog.show();
    }

    @Override
    protected Boolean doInBackground(Void... params) {
        // Download the JSON file. The URL is in params[0].
        BufferedReader bufferedReader = null;
        Boolean hazardsLoadedOK = Boolean.TRUE;

        try {
            if (ConfigSingleton.getInstance()
                    .loadIncidentsFromLocalJSONFile()) {
                String assetFileName = ConfigSingleton.getInstance()
                        .localIncidentsJSONFile();
                InputStream input = ctx.getAssets().open(assetFileName);
                int size = input.available();
                byte[] buffer = new byte[size];
                input.read(buffer);
                input.close();
                String text = new String(buffer);

                HazardDatabase.Companion.getInstance().init(text);
            } else {
                URL url = new URL(ConfigSingleton.getInstance()
                        .remoteIncidentsJSONFile());
                InputStreamReader instreamReader = new InputStreamReader(
                        url.openStream());
                bufferedReader = new BufferedReader(instreamReader);

                String line = null;
                StringBuffer lineBuffer = new StringBuffer();

                while ((line = bufferedReader.readLine()) != null) {
                    lineBuffer.append(line);
                }

                HazardDatabase.Companion.getInstance().init(lineBuffer.toString());
                bufferedReader.close();
            }
        } catch (Exception e) {
            MLog.w(TAG, "Failed to load hazards JSON", e);
            hazardsLoadedOK = Boolean.FALSE;
        } finally {
            if (bufferedReader != null) {
                try {
                    bufferedReader.close();
                } catch (IOException e) {
                }
            }
        }

        return hazardsLoadedOK;
    }

    @Override
    protected void onPostExecute(Boolean result) {
        // Force the listView to refresh its data

        if (dialog != null) {
            dialog.dismiss();
        }

        if (result == Boolean.TRUE) {
            listWithEmptyMessage.setAdapter(new HazardListAdapter(ctx, mode.getFilter()));
        } else {
            // We don't all mainLayout.setAdapter, which means that the old
            // (stale)
            // data will still remain visible.
            MLog.i(TAG,
                    "Failed to load camera image - showing error dialog");
            AlertDialog.Builder builder = new AlertDialog.Builder(ctx);
            builder.setTitle("No Incident Data");
            builder.setCancelable(true);
            builder.setMessage("Tap the refresh icon at top right to try again.");
            builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    dialog.cancel();
                }
            });
            AlertDialog dialog = builder.create();
            dialog.show();
        }
    }
}

File: app/src/main/java/rod/bailey/trafficatnsw/hazard/DownloadHazardFileTask.kt
package rod.bailey.trafficatnsw.hazard

import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Context
import android.os.AsyncTask
import rod.bailey.trafficatnsw.R
import rod.bailey.trafficatnsw.ui.view.ListWithEmptyMessage
import rod.bailey.trafficatnsw.util.ConfigSingleton
import rod.bailey.trafficatnsw.util.MLog
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.URL

/**
 * Async task for downloading the incident.json file from the Live Traffic web site
 */
class DownloadHazardFileTask(private val ctx: Context,
							 private val hazardList: ListWithEmptyMessage) : AsyncTask<Void, Void, Boolean>() {
	private var dialog: ProgressDialog? = null

	/**
	 * Show a "loading incidents..." progress dialog
	 */
	override fun onPreExecute() {
		super.onPreExecute()

		dialog = ProgressDialog(ctx)
		dialog?.setMessage(ctx.getString(R.string.hazards_list_load_progress_msg))
		dialog?.setCancelable(true)
		dialog?.isIndeterminate = true
		dialog?.show()
	}

	private fun loadIncidentsFromLocalJSONFile(): String? {
		MLog.d(LOG_TAG, "Loading incidents from local JSON file")
		val assetFileName = ConfigSingleton.instance.localIncidentsJSONFile()
		val input = ctx.assets.open(assetFileName)
		val size = input.available()
		val buffer = ByteArray(size)
		input.read(buffer)
		input.close()
		val text = String(buffer)
		return text
	}

	private fun loadIncidentsFromRemoteJSONFile(): String? {
		var bufferedReader: BufferedReader? = null
		var result: String? = null

		try {
			MLog.d(LOG_TAG, "Loading incidents from remote JSON file")
			val url = URL(ConfigSingleton.instance.remoteIncidentsJSONFile())
			val inStreamReader = InputStreamReader(url.openStream())
			bufferedReader = BufferedReader(inStreamReader)
			var line: String?
			val lineBuffer = StringBuffer()
			var eof: Boolean = false

			while (!eof) {
				line = bufferedReader.readLine()
				if (line == null) {
					eof = true
				} else {
					lineBuffer.append(line)
				}
			}

			bufferedReader.close()
			result = lineBuffer.toString()
		}
		catch (e: Throwable) {
			MLog.w(LOG_TAG, "Failed to load hazards JSON", e)
		}
		finally {
			try {
				bufferedReader?.close()
			}
			catch (e: IOException) {
				MLog.w(LOG_TAG, e)
			}
		}

		return result
	}

	override fun doInBackground(vararg params: Void): Boolean {
		val jsonText: String? =
			if (ConfigSingleton.instance.loadIncidentsFromLocalJSONFile()) {
				loadIncidentsFromLocalJSONFile()
			} else {
				loadIncidentsFromRemoteJSONFile()
			}

		if (jsonText != null) {
			HazardCacheSingleton.instance.init(jsonText)
			return true
		}
		return false
	}

	override fun onPostExecute(result: Boolean) {
		// Hide the progress dialog
		dialog?.dismiss()

		if (result) {
			hazardList.setAdapter(HazardListAdapter())
		} else {
			// We haven't called mainLayout.setAdapter, which means that the old
			// (stale) data will still remain visible under this error dialog
			MLog.i(LOG_TAG, "Failed to load hazard JSON - showing error dialog")
			val builder = AlertDialog.Builder(ctx)
			builder.setTitle(ctx.getString(R.string.hazards_list_load_failure_dialog_title))
			builder.setCancelable(true)
			builder.setMessage(ctx.getString(R.string.hazards_list_load_failure_dialog_msg))
			builder.setPositiveButton(
				ctx.getString(R.string.hazards_list_load_failure_dialog_positive_button),
				{ dialog, _ -> dialog.cancel() })
			builder.create().show()
		}
	}

	companion object {
		private val LOG_TAG = DownloadHazardFileTask::class.java.simpleName
	}
}
